toolchain/sem_ir/inst_kind.h

```diff
@@ -16,14 +16,37 @@ namespace Carbon::SemIR {
 // of value.
 enum class InstValueKind : int8_t {
   // This instruction doesn't produce a value, and shouldn't be referenced by
-  // other
-  // instructions.
+  // other instructions.
   None,
   // This instruction represents an expression or expression-like construct that
   // produces a value of the type indicated by its `type_id` field.
   Typed,
 };
 
+// Whether an instruction can be used to define a constant value. This specifies
+// whether the instruction can be added to the `constants()` list. Note that
+// even instructions that cannot define a constant value can still have an
+// associated `constant_value()`, but the constant value will be a different
+// kind of instruction.
+enum class InstConstantKind : int8_t {
+  // This instruction never defines a constant value. For example,
+  // `UnaryOperatorNot` never defines a constant value; if its operand is a
+  // template constant, its constant value will instead be a `BoolLiteral`. This
```

Comment:
Seems a little odd, structurally, that a BoolLiteral constant isn't a template constant; but maybe I'm missing the purpose...

Reply:
That's reversed from what the comment is trying to say: a `BoolLiteral` can be a constant value; a `UnaryOperatorNot` cannot, because evaluating a `UnaryOperatorNot` produces a `BoolLiteral` constant not a `UnaryOperatorNot` constant.